GVFS/GVFS.Common/ModifiedPathsDatabase.cs

```diff
@@ -52,6 +52,34 @@ public static bool TryLoadOrCreate(ITracer tracer, string dataDirectory, Physica
             return true;
         }
 
+        public void Compress(ITracer tracer)
+        {
+            int startingCount = this.modifiedPaths.Count;
+            using (ITracer activity = tracer.StartActivity("Compress ModifiedPaths", EventLevel.Informational))
+            {
+                foreach (var item in this.modifiedPaths)
+                {
+                    int pathSeparatorIndex = item.IndexOf('/');
+                    while (pathSeparatorIndex >= 0 && pathSeparatorIndex < item.Length - 1)
+                    {
+                        string folder = item.Substring(0, pathSeparatorIndex + 1);

```

Comment:
This is another subtle detail. The `+ 1` here is super important, to ensure that there is a trailing `/`, and avoid matching on a file with the same name rather than a folder with that name.

I would make this much more explicit by removing the `+ 1`, and adding a `+ "/"` at the end. But this issue will be moot if you use `string.Split`, since the `/`'s won't be part of the strings, and you'll have to forcibly reinsert them anyway.